scrapyd/webservice.py

```diff
@@ -1,67 +1,97 @@
-from copy import copy
+import functools
+import re
 import traceback
 import uuid
-try:
-    from cStringIO import StringIO as BytesIO
-except ImportError:
-    from io import BytesIO
+from copy import copy
+from io import BytesIO
 
 from twisted.python import log
+from twisted.web.error import Error
+from twisted.web.http import Request
+
+from .utils import get_spider_list, JsonResource, UtilsCache, \
+    native_stringify_dict
+
+
+def with_safe_project_name(func):
+    @functools.wraps(func)
+    def wrapper(resource, txrequest):
+        project_name = txrequest.args.pop(b'project', [None])[0]
+        msg = "'Project' name is required and must not contain illegal characters. "
+        msg += f'Project name "{project_name}" is not valid'
+        if not project_name:
+            raise Error(code=400, message=msg.encode())
+
+        project_name = project_name.decode()
+
+        allowed_name = re.sub('[^A-Za-z0-9.]+', '-', project_name)
```

Comment:
As @Digenis wrote in #117, this will break backwards compatibility. If I have an existing project name that contains `@` or `_` or `-` or any number of other legal, secure characters, then I can no longer interact with that project using Scrapyd's API. We can maybe make it "opt-in" behavior, and change the default [configuration file](https://scrapyd.readthedocs.io/en/stable/config.html) to opt-in to this behavior. That way, old deployments that don't have the setting in the conf file will not perform this sanitization.

Reply:
> we can maybe make it "opt-in" behavior, and change the default configuration file to opt-in to this behavior. 

the goal of this is to secure project from doing something like POST file to /etc/hosts or worse deleteversion /etc/hosts. We should not make this opt-in, should be built in. There is always an option to override scrapyd resources, and adding your handler in scrapyd configuration, you can for example add your own addversion.json. 

> If I have an existing project name that contains @ or _ or - or any number of other legal, secure characters, then I can no longer interact with that project using Scrapyd's API.

yeah you're right, after thinking about this, maybe the approach in this commit is too aggressive, and can break things for some valid use cases. Maybe we could release a milder version in 1.3 and add some deprecation warning saying that in next version, say 2.0. Project names will not accept some characters? Now question which characters to allow in next release, I would say probably all characters that are ok in setuptools.setup(name=) function? So for instance `@` won't be allowed, but space should be allowed? 

To sum up I'll change this to disallow only insecure characters, allow some legal secure characters. Let me know what you think.

Reply:
Or maybe let's do it differently, let's add project name sanitization in eggstorage in this PR. Fix project filename there, like we fix version. And in web service just print some deprecation warning in logs about invalid names

Reply:
My thinking was to make it opt-in for **existing** projects in 1.x, opt-out (i.e. built-in) for **new** projects in 1.x, and opt-out for **all** projects in 2.x. This could be done by having a configuration setting whose default value is False in 1.x, but True in 2.x. The default template for the configuration file would set the value to True, such that existing projects (whose configuration file doesn't set a value) would use the default value of False, but new projects would have a configuration file that sets it to True.

That said, if we can relax the validation to avoid breaking changes (your first suggestion), or if we can do something else that progressively changes the behavior (if I understand your second suggestion), that's fine too.

Reply:
> That said, if we can relax the validation to avoid breaking changes (your first suggestion), or if we can do something else that progressively changes the behavior (if I understand your second suggestion), that's fine too.

I would prefer progressive enhancement, but there is no way to completely avoid breaking it for some people here. If someone was storing projects with unusual characters, it is always begging for trouble. People who have projects "some/project" will have problem. But if anyone is actually doing something like this, it's on their own risk. 

I think we can trust users to have some knowledge and experience. If someone can install and update ScrapyD he should also be able to read documentation, revert to previous version, remove eggs with bad filenames and deploy new ones using API. 

 We will also give warning in docs about it.

Reply:
> That said, if we can relax the validation to avoid breaking changes (your first suggestion)

I relaxed validation here https://github.com/scrapy/scrapyd/pull/421/commits/49ab5a77ce5690f5dec1caccfd337c4bd2234c1a, but it is not completely avoiding breaking changes. Still, I think there is no way to avoid it, if someone uses project name "@my project@" he should really prepare for trouble. 

> My thinking was to make it opt-in for existing projects in 1.x, opt-out (i.e. built-in) for new projects in 1.x, and opt-out for all projects in 2.x. This could be done by having a configuration setting whose default value is False in 1.x, but True in 2.x. 

I thought about it again, it is not a bad idea, but it would add complexity to the project and increase development time and future maintenance time. But I thought about something else, how about we just launch 2.0 in 2022 and just announce this and other changes? 2.0 gives some room for breaking changes and people will tolerate them. It won't be a big 2.0 I guess, because nothing big will change, but it would justify making backward incompatible change here.

Reply:
2.0 instead of 1.3 is fine with me! I'm not too precious about the size of releases :)